src/compiler/checker.ts

```diff
@@ -3643,9 +3630,17 @@ module ts {
         }
 
         function getReducedTypeOfUnionType(type: UnionType): Type {
-            // If union type was created without subtype reduction, perform the deferred reduction now
+            // If union type was created without subtype reduction, perform the deferred reduction now. If a circularity
+            // is detected, simply use the type itself.
             if (!type.reducedType) {
-                type.reducedType = getUnionType(type.types, /*noSubtypeReduction*/ false);
+                type.reducedType = resolvingType;
+                let reducedType = getUnionType(type.types, /*noSubtypeReduction*/ false);
+                if (type.reducedType === resolvingType) {
+                    type.reducedType = reducedType;
+                }
+            }
+            else if (type.reducedType === resolvingType) {
+                type.reducedType = type;

```

Comment:
Subtype reduction is basically an optimization we do to avoid excessively large union types, which take longer to process and look strange in quick info and error messages. Semantically there is no difference between the reduced type and the type itself. So, when we detect a circularity we simply say that the reduced type is the type itself.